### PR TITLE
ARROW-9768: [Python] Check overflow in conversion of datetime objects to nanosecond timestamps

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
In principle, we have a `safe=True/False` keyword in `pa.array(..)` that could be honored here, and only do the overflow check if `safe=True` (but we currently don't discriminate for safe=True/False anywhere in python_to_arrow.cc)

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 1]
```diff
@@ -263,7 +264,12 @@ struct ValueConverter<TimestampType> {
           value = internal::PyDateTime_to_us(dt) - offset * 1000 * 1000;
           break;
         case TimeUnit::NANO:
-          value = internal::PyDateTime_to_ns(dt) - offset * 1000 * 1000 * 1000;
+          // Conversion to nanoseconds can overflow -> check multiply of microseconds
+          value = internal::PyDateTime_to_us(dt);
+          if (arrow::internal::MultiplyWithOverflow(value, 1000, &value)) {
+            return internal::InvalidValue(obj, "out of bounds for nanosecond resolution");
+          }
+          value -= offset * 1000 * 1000 * 1000;
```
Comment (pitrou) [inline cpp/src/arrow/python/python_to_arrow.cc:272]: This subtraction could overflow as well, no?

Comment (jorisvandenbossche) [reply cpp/src/arrow/python/python_to_arrow.cc:272]: The offset is always a small number (time zone offsets are always less then 12 hours ?), so shouldn't give any problem?

Comment (jorisvandenbossche) [reply cpp/src/arrow/python/python_to_arrow.cc:272]: Ah, the offset itself won't overflow, but for the *subtraction* (as you actually commented ..) that could indeed happen in a small corner case: if your naive timezone is close to the lower or upper limit of the nanosecond range, applying the offset could indeed overflow.

**Review by pitrou — APPROVED** [review]
+1, LGTM

**Comment by jorisvandenbossche** [comment]
@pitrou do you know why the windows macro thing comes up in this PR?

**Comment by pitrou** [comment]
@jorisvandenbossche Because `int_util_internal.h` indirectly includes some more Windows headers (through a vendored dependency).

**Comment by pitrou** [comment]
I will presume the s390x build is ok and merge.

**Comment by jorisvandenbossche** [comment]
> Because int_util_internal.h indirectly includes some more Windows headers (through a vendored dependency).

OK, thanks for fixing that!

**Pull request closed without merging**